galsim/config/extra.py

```diff
@@ -0,0 +1,377 @@
+                continue
+
+            extra_obj = GetFinalExtraOutput(key, config, logger)
+
+            # If we have a method, we need to attach it to the extra_obj, since it might
+            # be a proxy, in which case the method call won't work.
+            if inspect.ismethod(write_func):
+                write_func = eval('extra_obj.' + write_func.__name__)
+                args = (file_name,)
+            else:
+                args = (extra_obj, file_name)
+            galsim.config.RetryIO(write_func, args, ntries, file_name, logger)
+            config['extra_objs_last_file'][key] = file_name
+            if logger and logger.isEnabledFor(logging.DEBUG):
+                logger.debug('file %d: Wrote %s to %r',config['file_num'],key,file_name)
+
+
+def BuildExtraOutputHDUs(config, logger=None, first=1):
+    """Write the extra output objects to either HDUS or images as appropriate.
+
+    This gets run at the end of the functions for building the regular output files.
+
+    Note: the extra items must have hdu numbers ranging continuously (in any order) starting
+    at first.  Typically first = 1, since the main image is the primary HDU, numbered 0.
+
+    @param config       The configuration dict.
+    @param logger       If given, a logger object to log progress. [default: None]
+    @param first        The first number allowed for the extra hdus. [default: 1]
+
+    @returns a list of HDUs and/or Images to put in the output FITS file.
+    """
+    config['index_key'] = 'file_num'
+    if 'output' in config:
+        output = config['output']
+        hdus = {}
+        for key in [ k for k in valid_extra_outputs.keys() if k in output ]:
+            hdu_func = valid_extra_outputs[key]['hdu']
+            if hdu_func is None: continue
+
+            field = output[key]
+            if 'hdu' in field:
+                hdu = galsim.config.ParseValue(field,'hdu',config,int)[0]
+            else:
+                # If no hdu, then probably writing to file
+                continue
+            if hdu <= 0 or hdu in hdus.keys():
+                raise ValueError("%s hdu = %d is invalid or a duplicate."%hdu)
+
+            extra_obj = GetFinalExtraOutput(key, config, logger)
+
+            # If we have a method, we need to attach it to the extra_obj, since it might
+            # be a proxy, in which case the method call won't work.
+            if inspect.ismethod(hdu_func):
+                hdu_func = eval('extra_obj.' + hdu_func.__name__)
+                hdus[hdu] = hdu_func()
+            else:
+                hdus[hdu] = hdu_func(extra_obj)
+
+        for h in range(first,len(hdus)+first):
+            if h not in hdus.keys():
+                raise ValueError("Cannot skip hdus.  Not output found for hdu %d"%h)
+        # Turn hdus into a list (in order)
+        hdulist = [ hdus[k] for k in range(first,len(hdus)+first) ]
+        return hdulist
+    else:
+        return []
+
+valid_extra_outputs = {}
```

Comment:
Maybe move this to the top of the file?  I was a bit confused by some logic above that referenced it before I realized it was a module-level variable.